src/lib/server/ring_buffer.hpp

```diff
@@ -0,0 +1,182 @@
+class RingBuffer {
+ public:
+  RingBuffer() = default;
+
+  char* data() noexcept { return _data.begin(); }
+
+  // Problem: full and empty might be same state, so head == tail
+  // Solution: Full state is tail + 1 == head
+  //           Empty state is head == tail
+  size_t size() const {
+    const auto current_size = _current_position - _start_position;
+    if (current_size < 0) {
+      return current_size + BUFFER_SIZE;
+    } else {
+      return current_size;
+    }
+  }
+
+  // See comment above
+  size_t maximum_capacity() const { return BUFFER_SIZE - 1; }
+
+  bool full() const { return size() == maximum_capacity(); }
+
+ protected:
+  std::array<char, BUFFER_SIZE> _data;
+  RingBufferIterator _start_position = RingBufferIterator(_data);
+  RingBufferIterator _current_position = RingBufferIterator(_data);
+};
+
+// Dedicated buffer for read operations. The ring buffer gets extended by methods for reading different data types.
+template <typename SocketType>
+class ReadBuffer : public RingBuffer {
+ public:
+  explicit ReadBuffer(std::shared_ptr<SocketType> socket) : _socket(socket) {}
+
+  // Extract numerical values from buffer. Values will be converted into the correct byte order if necessary.
+  template <typename T, typename std::enable_if_t<std::is_same_v<uint16_t, T> || std::is_same_v<int16_t, T>, int> = 0>
+  T get_value() {
+    _receive_if_necessary(sizeof(T));
+    T network_value = 0;
+    std::copy_n(_start_position, sizeof(T), reinterpret_cast<char*>(&network_value));
+    _start_position += sizeof(T);
+    return ntohs(network_value);
+  }
+
+  template <typename T, typename std::enable_if_t<std::is_same_v<uint32_t, T> || std::is_same_v<int32_t, T>, int> = 0>
+  T get_value() {
+    _receive_if_necessary(sizeof(T));
+    T network_value = 0;
+    std::copy_n(_start_position, sizeof(T), reinterpret_cast<char*>(&network_value));
+    _start_position += sizeof(T);
+    return ntohl(network_value);
+  }
+
+  template <typename T, typename std::enable_if_t<!std::is_same_v<uint16_t, T> && !std::is_same_v<int16_t, T> &&
+                                                      !std::is_same_v<uint32_t, T> && !std::is_same_v<int32_t, T>,
+                                                  int> = 0>
+  T get_value() {
+    _receive_if_necessary(sizeof(T));
+    T network_value = 0;
+    std::copy_n(_start_position, sizeof(T), reinterpret_cast<char*>(&network_value));
+    _start_position += sizeof(T);
+    return network_value;
+  }
+
+  // String functions
+  std::string get_string(const size_t string_length, const bool has_null_terminator = true);
+  std::string get_string();
+
+  PostgresMessageType get_message_type();
+
+ private:
+  std::shared_ptr<SocketType> _socket;
+  void _receive_if_necessary(const size_t bytes_required = 1);
+};
+
+// Dedicated buffer for write operations. The ring buffer gets extended by methods for writing different data types.
+template <typename SocketType>
+class WriteBuffer : public RingBuffer {
```

Comment:
This is not inheritance. The WriteBuffer is not a special type of a RingBuffer, but does something completely different (talking to boost asio)